### PR TITLE
Fix issue 132: size the worker-specific vectors correctly.

### PR DESCRIPTION
**Pull request by reunanen**
Fix issue 132: size the worker-specific vectors correctly.

### CODE REVIEW

File: tiny_cnn/layers/layer.h [hunk 1]
```diff
@@ -238,55 +240,79 @@ class layer_base {
...
+    }
+
 protected:
     cnn_size_t in_size_;
     cnn_size_t out_size_;
     bool parallelize_;
 
     layer_base* next_;
     layer_base* prev_;
-    vec_t a_[CNN_TASK_SIZE];          // w * x
```
Comment (reunanen) [inline tiny_cnn/layers/layer.h]: This is the beef of the pull request. These should not be fixed-size, unless n_threads given to train() is somehow limited. (Currently it isn't, so we may overflow these arrays – in particular when not using OpenMP, in which case `CNN_TASK_SIZE` is only 8.)

File: tiny_cnn/layers/convolutional_layer.h [hunk 2]
```diff
@@ -489,15 +493,21 @@ class convolutional_layer : public layer<Activation> {
...
+        conv_layer_worker_storage_.resize(worker_count);
+        init();
+    }
+
 private:
     void init() {
-        for (cnn_size_t i = 0; i < CNN_TASK_SIZE; i++) {
+        for (conv_layer_worker_specific_storage& cws : conv_layer_worker_storage_) {
             if (pad_type_ == padding::same) {
-                prev_out_buf_[i] = new vec_t(in_padded_.size(), float_t(0));
```
Comment (reunanen) [inline tiny_cnn/layers/convolutional_layer.h]: I wonder if there was a memory leak here. Looks like the allocated `prev_out_buf_` was never `free`d?

Anyway fixed now by avoiding manual memory management.

File: tiny_cnn/layers/dropout_layer.h [hunk 3]
```diff
@@ -43,45 +43,39 @@ class dropout_layer : public layer<activation::identity> {
           dropout_rate_(dropout_rate),
           scale_(float_t(1) / (float_t(1) - dropout_rate_))
     {
-        mask_ = new bool[in_size_ * CNN_TASK_SIZE];
```
Comment (reunanen) [inline tiny_cnn/layers/dropout_layer.h]: Avoiding manual memory management here as well.

**Comment by pansk** [comment]
Thank you for your clarification, I think we got the point: n_threads is supposed to be in the range [1, CNN_TASK_SIZE]. If it's not enforced (as it seems), that's where the bug is (all layer classes work like that, if we decide to make it dynamic, we should change it everywhere).
The allocated prev_out_buf_ could be memory leak, I'll look at it.

**Comment by reunanen** [comment]
@pansk : Yes, well, I tried to change it (to be dynamic) pretty much everywhere, though it's quite possible that I missed something.

**Comment by pansk** [comment]
@reunanen My fault, I saw few changes, but they were comments. Apparently it's complete: :+1: for me :)
I don't know @nyanp 's opinion on this, but from my point of view this could be a good improvement (and allow the user to have a better control over allocated memory). I'd just differentiate (if such a distinction makes sense) memory allocated for inference-only from that used by training, and add a function to it to release all the training-related memory, and to set up an inference threading level of N (with default to one), in order to have a slim single-threaded inference-only object. I can take care of this if the PR goes through.

**Comment by reunanen** [comment]
@pansk : Yep, there were so many changes that I thought it'd be best to highlight (at least to some extent) what is important and what is boilerplate.

**Comment by nyanp** [comment]
@reunanen 
Thanks, I think it's a great improvement! I've wanted to shrink memory footprint after training phase (memory footprint is always important for embedding systems), and your PR is very useful for it :+1: 

Encapsulation weights, diffs as `storage` will probably be a good first step for further improvement such as weights sharing over layers (it is necessary for lstm), and gpu supporting.

**Pull request merged**